### PR TITLE
Separate ldns parsing from clap parsing

### PR DESCRIPTION
**Pull request by tertsdiepraam**
This PR adds a separate code path for parsing arguments in a ldns-compatible way. The ldns tools cannot (in general) be parsed with clap, so they need full flexibility over how they parse their arguments. This parsing is only used when the binary is executed with the name of an ldns tool (either via renaming or a symlink). This also allows us to provide different defaults for the ldns and dnst tools.

### CODE REVIEW

**Review by ximon18 — APPROVED** [review]
LGTM. Added some minor comments.

**Comment by ximon18** [comment]
One possible downside I see of this PR thinking about it a bit late (as I just submitted a review), is that this approach requires all implemented LDNS commands to also be implemented for `dnst`, as the final value produced by the alternate code is of type `Args`. I'm not sure that we will want to support the same set of commands for `dnst` as we implement for backward compatibility with LDNS.

File: src/main.rs
Comment (tertsdiepraam) [reply src/main.rs:44]: Rationale was to do the same as ldns until we have figured out what we want for `dnst`. I'm happy with 2 for `dnst` as well.

File: src/main.rs
Comment (ximon18) [reply src/main.rs:17]: No, just a better error message than `unwrap()` will produce.

File: src/commands/mod.rs
Comment (ximon18) [reply src/commands/mod.rs:68]: Ah I missed that.

File: src/commands/nsec3hash.rs
Comment (tertsdiepraam) [reply src/commands/nsec3hash.rs:16]: I believe that this should be default behaviour for all CLI tools so that it's more convenient to create aliases and override the behaviour of an alias. Not really an issue for this tool, but I think we should just enable it on all tools.

Comment (ximon18) [reply src/commands/nsec3hash.rs:16]: I'm not sure to what you are referring regarding aliases. Isn't it confusing to not error out when an arg that is expected once is given more than once?

**Comment by tertsdiepraam** [comment]
> One possible downside I see of this PR thinking about it a bit late (as I just submitted a review), is that this approach requires all implemented LDNS commands to also be implemented for dnst, as the final value produced by the alternate code is of type Args. I'm not sure that we will want to offer dnst equivalents for all implemented LDNS commands.

Good point. I'll check if we can work around that with `clap`.

File: src/main.rs
Comment (tertsdiepraam) [reply src/main.rs]: I'm returning early now if the `next`, `file_name` or `to_str` calls fail.

File: src/commands/nsec3hash.rs
Comment (tertsdiepraam) [reply src/commands/nsec3hash.rs:16]: As an example of overriding aliases, with `ls`, you can do this:
```bash
alias my_cool_ls="ls --color --hyperlink --human-readable --classify"
```
but then if you want all those options but not the color bit, you can do
```
my_cool_ls --color=none
```
While keeping all the other fancy options.

File: src/commands/mod.rs
Comment (ximon18) [inline src/commands/mod.rs:65]: I wonder could this be improved to know what the last arg parsed by `parser.next()` was? To save boiler plate repeating the argument letter?

File: src/commands/mod.rs
Comment (bal-e) [inline src/commands/mod.rs:77]: I don't think these should be defined in the trait, but rather at the module level.

File: src/commands/nsec3hash.rs
Comment (bal-e) [reply src/commands/nsec3hash.rs:16]: I'm not particularly swayed either way.  I like the convenience of the aliases but I don't think the ldns commands would be used that way and I agree that it can cause some confusion, particularly when hidden away in a script somewhere.

**Review by bal-e — COMMENTED** [review]
I like this solution to ldns compatibility!  I only have one actual comment, after that LGTM.

**Comment by tertsdiepraam** [comment]
Re `args_override_self`: this PR was not the right place to introduce it anyway. I'll remove it.

File: src/commands/mod.rs
Comment (tertsdiepraam) [reply src/commands/mod.rs:77]: Yeah that does make sense. I just didn't find a good place to put it. I'll give it another try.

File: src/commands/nsec3hash.rs
Comment (tertsdiepraam) [reply src/commands/nsec3hash.rs]: Readability? We could just do `-a`, I don't really mind.

File: Cargo.toml
Comment (tertsdiepraam) [reply Cargo.toml:9]: I'm not going to do this now. I think the code is actually quite subtle (especially with invalid utf-8 and windows support and all that).

Comment (ximon18) [reply Cargo.toml:9]: Indeed it doesn't have to be now, but prior to release we might want to consider vendoring `lexopt` as is.

File: src/commands/nsec3hash.rs
Comment (ximon18) [reply src/commands/nsec3hash.rs]: I would go for just `-X` and ideally also find a way to simplify this case of `Arg::Short('X') => { ... "-X" }` as its error prone, in testing this I already accidentally once did `Arg::Short('X') => { ... "-Y" }`...

Comment (ximon18) [reply src/commands/nsec3hash.rs]: I'm mainly in favour of the shorter text to avoid more duplication, but that's also why I'd be interested in seeing (not necessarily now) if we can automate this duplication to avoid accidentally using different values when they should be the same. For me we can merge this PR without resolving this.

**Review by ximon18 — APPROVED** [review]

File: src/commands/mod.rs
Comment (ximon18) [reply src/commands/mod.rs]: This could be improved later, it shouldn't block this PR.

**Review by bal-e — APPROVED** [review]

**Pull request merged**
